astropy/nddata/covariance.py

```diff
@@ -0,0 +1,1034 @@
+            raise ValueError("Fewer than two samples provided!")
+        return Covariance.from_array(np.cov(samples), cov_tol=cov_tol, rho_tol=rho_tol)
+
+    @classmethod
+    def from_array(cls, covar, cov_tol=None, rho_tol=None, raw_shape=None):
+        r"""
+        Define a covariance object using a dense array.
+
+        Parameters
+        ----------
+        covar : array-like
+            Array with the covariance data. The shape of the array must be
+            square. Input can be any object that can be converted to a dense
+            array using the object method ``toarray`` or using
+            ``numpy.atleast_2d``.
+
+        cov_tol : :obj:`float`, optional
+            Any covariance value less than this is assumed to be equivalent to
+            (and set to) 0.
+
+        rho_tol : :obj:`float`, optional
+            Any correlation coefficient less than this is assumed to be
+            equivalent to (and set to) 0.
+
+        raw_shape : :obj:`tuple`, optional
+            The covariance data is for a higher dimensional array with this
+            shape.  For example, if the covariance data is for a 2D image with
+            shape ``(nx,ny)`` -- the shape of the covariance array is be
+            ``(nx*ny, nx*ny)`` -- set ``raw_shape=(nx,ny)``. This is primarily
+            used for reading and writing; see also :func:`transpose_raw_shape`.
+            If None, any higher dimensionality is ignored.  Currently, this is
+            only allowed if the raw shape is 2D.
+
+        Returns
+        -------
+        :class:`Covariance`
+            The covariance matrix built using the provided array.
+
+        Raises
+        ------
+        ValueError
+            Raised if ``covar`` could not be converted to a dense array.
+        """
+        try:
+            _covar = covar.toarray()
+        except AttributeError:
+            _covar = np.atleast_2d(covar)
+        if not isinstance(_covar, np.ndarray) or _covar.ndim != 2:
+            raise ValueError(
+                "Could not convert input covariance data into a 2D dense array."
+            )
+
+        n = _covar.shape[0]
+        if rho_tol is not None:
+            variance = np.diag(_covar)
+            correlation = _covar / np.ma.sqrt(variance[:, None] * variance[None, :])
+            correlation[np.ma.absolute(correlation) < rho_tol] = 0.0
+            _covar = correlation.filled(0.0) * np.ma.sqrt(
+                variance[:, None] * variance[None, :]
+            ).filled(0.0)
+        if cov_tol is not None:
+            _covar[_covar < cov_tol] = 0.0
+
+        indx = _covar > 0.0
+        i, j = np.meshgrid(np.arange(n), np.arange(n), indexing="ij")
+        return cls(
+            sparse.coo_matrix(
+                (_covar[indx].ravel(), (i[indx].ravel(), j[indx].ravel())), shape=(n, n)
+            ).tocsr(),
+            impose_triu=True,
+            raw_shape=raw_shape,
+        )
+
+    @classmethod
+    def from_fits(
```

Comment:
This method relies too much on special FITS header keywords. If these keywords only apply to, say, one very specific telescope or instrument, it should not be in core `astropy` library.

Reply:
Are you specifically referring to the `INDX*` and `RHO*` keywords?  Yes, these are specific to this class, but they are not specific to any instrument or telescope.  They are consistent with general nomenclature used for covariance matrices.  Although, I should probably generalize the `I` and `J` to something like `0` and `1`...

Reply:
So, these are formalized in some WCS papers or something? (I never read them LoL)

Reply:
No, there's no formal WCS for these (that I know of).